core/templates/services/html-length.service.ts

```diff
@@ -43,58 +61,114 @@ export class HtmlLengthService {
+          'that a valid string that includes HTML tags is provided.');
+      }
+    } catch (error) {
+      throw new Error(
+        'Failed to parse HTML string.' +
+        ' Ensure that a valid string that includes HTML tags is provided.');
+    }
     let totalWeight = 0;
     for (let tag of Array.from(dom.body.children)) {
-      const ltag = tag.tagName.toLowerCase();
-      if (this.textTags.includes(ltag)) {
-        totalWeight += this.getWeightForTextNodes(tag as HTMLElement);
+      const lowercaseTagName = tag.tagName.toLowerCase();
+      if (this.textTags.includes(lowercaseTagName)) {
+        totalWeight +=
+        this.getWeightForTextNodes(
+          tag as HTMLElement, calculationType);
       }
     }
     return totalWeight;
   }
 
-  private getWeightForTextNodes(textNode: HTMLElement): number {
+  private getWeightForTextNodes(
+      textNode: HTMLElement, calculationType: CalculationType): number {
     const textContent = textNode.textContent || '';
-    const words = textContent.trim().split(' ');
-    const wordCount = words.length;
-
-    return wordCount;
+    return this.calculateTextWeight(textContent, calculationType);
   }
 
-  private getWeightForNonTextNodes(nonTextNode: string): number {
-    if (nonTextNode.includes('oppia-noninteractive-math')) {
-      return 1;
+  private calculateTextWeight(
+      textContent: string, calculationType: CalculationType): number {
+    let trimmedTextContent = textContent.trim();
+    let totalWeight = 0;
+    if (calculationType === CALCULATION_TYPE_WORD && trimmedTextContent) {
+      const words = trimmedTextContent.split(' ');
+      totalWeight = words.length;
+    } else {
+      totalWeight = trimmedTextContent.length;
     }
-    const altTextMatch = nonTextNode.match(/alt-with-value=["']([^"']*)["']/);
-    let words = [];
-    if (altTextMatch && altTextMatch[1]) {
-      const altText = altTextMatch[1];
-      words = altText.trim().split(' ');
+    return totalWeight;
+  }
+
+  /* TODO(#19729): Create RTE-component-specific logic
+  for calculating the lengths of RTE-components */
+  getWeightForNonTextNodes(
+      nonTextNode: string, calculationType: CalculationType): number {
+    nonTextNode = this.htmlEscaperService.escapedStrToUnescapedStr(nonTextNode);
+    let domparser = new DOMParser();
+    let dom: Document;
+    dom = domparser.parseFromString(nonTextNode, 'text/html');
+    let domTag = dom.body.children[0];
+    let domId = domTag.tagName.toLowerCase();
+    switch (domId) {
+      case 'oppia-noninteractive-math':
+        return 1;
+      case 'oppia-noninteractive-collapsible':
+      case 'oppia-noninteractive-tabs':
+        return 1000;
+      case 'oppia-noninteractive-video':
+        return 0;
+      case 'oppia-noninteractive-link':
+      case 'oppia-noninteractive-skillreview': {
+        const textValueAttr = domTag.getAttribute('text-with-value');
+        const textValue = textValueAttr ? textValueAttr.replace(/"/g, '') : '';
```

Comment:
Something is a bit off here. When we use these tags in practice how do we do the unescaping to get the actual underlying value of the attributes? We should be able to follow the same process here and not resort to regex matching.

See if you can provide a link to the code that parses these tags in practice.

Reply:
If the hint has an image then you'd need to extract its alt text to count characters etc. right?

Btw, when I say "use these tags in practice", that means "use these tags in the app", not "use them in practice sessions".

Reply:
- We are extracting altText in next case, you might have missed that.
- Oh okay now I got what you meant with practice, so analysis done shows,
  - In ``core/templates/pages/contributor-dashboard-page/modal-templates/translation-modal.component.ts``  file also we are extracting altText for a image but its a more complex version what I did,  below is code reference of that file, In this they are extracting the value of alt-text value by using the substring method which is similar to slice(), which I have used as in now.
```
getElementAttributeTexts(
      elements: HTMLCollectionOf<Element>, type: string): string[] {
    const textWrapperLength = 6;
    const attributes = Array.from(elements, function(element: Element) {
      // A sample element would be as <oppia-noninteractive-image alt-with-value
      // ="&amp;quot;Image description&amp;quot;" caption-with-value=
      // "&amp;quot;Image caption&amp;quot;" filepath-with-value="&amp;quot;
      // img_2021029_210552_zbmdt94_height_54_width_490.png&amp;quot;">
      // </oppia-noninteractive-image>
      if (element.localName === 'oppia-noninteractive-image') {
        const attribute = element.attributes[
          type as keyof NamedNodeMap] as Attr;
        const attributeValue = attribute.value;
        return attributeValue.substring(
          textWrapperLength, attributeValue.length - textWrapperLength);
      }
    });
```

Reply:
Actually, the code sample you quoted isn't done correctly either, I'm afraid, because it uses substrings as well. Take a look at e.g. extensions/rich_text_components/image/directives/oppia-noninteractive-image.component.ts instead and see how htmlEscaperService is used -- that is the correct approach.